libs/clist/clist.go

```diff
@@ -115,43 +115,42 @@ func (e *CElement) Next() *CElement {
 // Nonblocking, may return nil if at the end.
 func (e *CElement) Prev() *CElement {
 	e.mtx.RLock()
-	defer e.mtx.RUnlock()
-
-	return e.prev
+	prev := e.prev
+	e.mtx.RUnlock()
+	return prev
 }
 
 func (e *CElement) Removed() bool {
 	e.mtx.RLock()
-	defer e.mtx.RUnlock()
-
-	return e.removed
+	isRemoved := e.removed
+	e.mtx.RUnlock()
+	return isRemoved
 }
 
 func (e *CElement) DetachNext() {
-	if !e.Removed() {
+	e.mtx.Lock()
+	if !e.removed {
+		e.mtx.Unlock()
```

Comment:
should we even bother unlocking before `panic`?

Reply:
I'm not sure. It depends on if we plan on ever catching panics. My thinking was that this may come up in testing, and may just be annoying to have it lock. (i.e. `require.Panics(t, <invalid thing>)`)